apis/duck/v1/source_types.go

```diff
@@ -170,3 +171,50 @@ type SourceList struct {
 
 	Items []Source `json:"items"`
 }
+
+func (s *Source) Validate(ctx context.Context) *apis.FieldError {
+	if s == nil {
+		return nil
+	}
+	return ValidateSource(ctx, *s).ViaField(apis.CurrentField)
+}
+
+// ValidateSource validates Source.
+func ValidateSource(ctx context.Context, src Source) *apis.FieldError {
+	return src.Spec.Validate(ctx).ViaField("spec")
+}
+
+func (s *SourceSpec) Validate(ctx context.Context) *apis.FieldError {
+	return s.CloudEventOverrides.Validate(ctx).ViaField("ceOverrides")
+}
+
+func (ceOverrides *CloudEventOverrides) Validate(ctx context.Context) *apis.FieldError {
+	for key := range ceOverrides.Extensions {
+		if err := validateExtensionName(key); err != nil {
+			return err.ViaField("extensions")
+		}
+	}
+	return nil
+}
+
+func validateExtensionName(key string) *apis.FieldError {
+	if len(key) < 1 {
+		return apis.ErrInvalidKeyName(
+			key,
+			"",
+			"CloudEvents attribute names MUST NOT be empty",
+		)
+	}
+
+	for _, c := range key {
+		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
+			return apis.ErrInvalidKeyName(
+				key,
+				"",
+				"CloudEvents attribute names MUST consist of lower-case letters ('a' to 'z') or digits ('0' to '9') from the ASCII character set",
```

Comment:
it is not clear that this is the message knative would like to show users. maybe something more like : `keys are expected to be lower-case letters ('a' to 'z') or digits ('0' to '9')` The fact that this is a cloudevents rule does not matter to knative so much. It will confuse people.